pennylane/utils.py

```diff
@@ -433,3 +434,78 @@ def expand_vector(vector, original_wires, expanded_wires):
     expanded_tensor = np.moveaxis(expanded_tensor, original_indices, wire_indices)
 
     return expanded_tensor.reshape(2 ** M)
+
+
+class ExecutionCounter:
+    """Context in which calls to a QubitDevice's execute method are counted without performing the execution.
+
+    The class mocks the execute method, returning a dummy value of the correct dimensions.
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane.utils import ExecutionCounter
+        qml.enable_tape()
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with ExecutionCounter() as counter:
+            circuit(0.1)
+            circuit(0.5)
+
+        print(counter.counts) # 2
+
+        with ExecutionCounter() as counter_grad:
+            g = qml.grad(circuit)
+            g(0.1)
+
+        # parameter-shift rule uses 2+1 executions
+        # to compute the gradient
+        print(counter_grad.counts) # 3
+
+    """
+
+    def __init__(self):
+        self.counts = None
+        self.temp_original = None
+
+    def __enter__(self):
+        # remember the original method
+        self.temp_original = qml._qubit_device.QubitDevice.execute
+
+        # Mock qubit device's execute method.
+        # The side effect will define the output of the mock.
+        qml._qubit_device.QubitDevice.execute = Mock(side_effect=self.side_effect)
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+
+        # record the number of calls to the mocked function
+        self.counts = qml._qubit_device.QubitDevice.execute.call_count
```

Comment:
For devices with `analytic=False`, should this also be multiplied by the number of shots, since each shot would be a separate execution on the device?

Reply:
This is a bit tricky - most backends count evaluating the same circuit for n shots as "1 execution". Or there may be a threshold, like 8641 or so shots in qiskit that are counted as one. So for now the idea of 1 execution, 1 count is the cleanest we can get.

If the tool gets used to estimate pricing we need to provide this information. Or integrate it into the recorder, which could be a device method that gets overwritten by each device.

Reply:
You could add `execute.shot_count` in addition to `execute.call_count`? That would be quite cool

Reply:
Good idea! Can extend the class that way in future?